File: BOJ/java/10819.java
/*
 * 차이를 최대로
 * https://www.acmicpc.net/problem/10819
 */

import java.util.Arrays;
import java.util.Scanner;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] a = new int[n];
        int result = 0;

        for (int i = 0; i < n; i++) {
            a[i] = sc.nextInt();
        }

        Arrays.sort(a);

        do {
            int temp = 0;
            for (int i = 0; i < n - 1; i++) {
                temp += Math.abs(a[i] - a[i + 1]);
            }
            if (result < temp) {
                result = temp;
            }
        } while (nextPermutation(a));

        System.out.println(result);
    }

    static boolean nextPermutation(int[] a) {
        int lastIndex = a.length - 1;
        int i = lastIndex;
        while (i > 0 && a[i] <= a[i - 1]) {
            i--;
        }
        if (i == 0) return false;

        int j = lastIndex;
        while (j >= i && a[j] <= a[i - 1]) {
            j--;
        }
        swap(a, i - 1, j);

        j = lastIndex;
        while (i < j) swap(a, i++, j--);
        return true;
    }

    static void swap(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
}

File: BOJ/kotlin/10819.kt
import kotlin.math.abs
import kotlin.math.max

fun main() {
    val n = readLine()!!.toInt()
    val a = readLine()!!.split(" ").map { it.toInt() }.sorted().toMutableList()

    var result = 0
    do {
        var t = 0
        for (i in 0 until a.size - 1) {
            t += abs(a[i] - a[i + 1])
        }
        result = max(result, t)
    } while (nextPermutation(a))
    println(result)
}

fun nextPermutation(a: MutableList<Int>): Boolean {
    val lastIndex = a.size - 1
    var i = (1..lastIndex).findLast { a[it] > a[it - 1] } ?: return false
    var j = (i..lastIndex).findLast { a[it] > a[i - 1] } ?: return false
    a.swap(i - 1, j)
    j = lastIndex
    while (i < j) a.swap(i++, j--)
    return true
}

fun <T> MutableList<T>.swap(i: Int, j: Int) {
    val temp = this[i]
    this[i] = this[j]
    this[j] = temp
}